spec/02-integration/15-queues/01-shutdown_spec.lua

```diff
@@ -0,0 +1,82 @@
+local helpers    = require "spec.helpers"
+
+
+local HTTP_SERVER_PORT = helpers.get_available_port()
+
+
+for _, strategy in helpers.each_strategy() do
+  describe("queue graceful shutdown [#" .. strategy .. "]", function()
+    local proxy_client
+
+    lazy_setup(function()
+      local bp = helpers.get_db_utils(strategy, {
+        "routes",
+        "services",
+        "plugins",
+      })
+
+      local service = bp.services:insert{
+        protocol = "http",
+        host     = helpers.mock_upstream_host,
+        port     = helpers.mock_upstream_port,
+      }
+
+      local route = bp.routes:insert {
+        hosts   = { "shutdown.test" },
+        service = service
+      }
+
+      bp.plugins:insert {
+        route = { id = route.id },
+        name     = "http-log",
+        config   = {
+          http_endpoint = "http://127.0.0.1:" .. HTTP_SERVER_PORT,
```

Comment:
What if the endpoint is configured with a hostname? This is where we ran into issues with Kong 3.x making http requests in the shutdown phase and had to alter DNS handling.

Reply:
Thank you for pointing this out.  We discussed that we'll need to switch DNS to synchronous mode during shutdown, I'm going to test and investigate.

Reply:
@27ascii When you write you had to alter DNS handling, can you provide some more information?  In my testing, I can observe that the DNS client no longer works when invoked while timers are run down.  When the DNS client can resolve the hostname from its cache, however, things work.  I wonder whether you had already found a way to make it continue to resolve even when shutting down.

Reply:
@hanshuebner With the hostname in the cache it works, however it's undetermined when it will expire. In order to have deterministic behaviour the DNS synchronization is disabled in the shutdown phase:
https://github.com/Kong/lua-resty-dns-client/blob/fb1e686a8857c7492fdd30ca228d200afbde19aa/src/resty/dns/client.lua#L433-L437

With this change we don't run into the issue where the DNS request times out.

Our change basically comes down to his: 
https://github.com/27ascii/kong-shutdown/blob/938b8b40609aa9dddbb6f1e68182b5c2ad163576/shutdown-events/handler.lua#L15-L17

Reply:
Thanks for this, I will see whether it can be arranged that DNS is switched to `dns_no_sync` mode automatically when shutting down.

Reply:
I've added https://github.com/Kong/kong/pull/10172/commits/c5f93b034c41affade463cfe2f96c397e96cc57b to implement this.  Feedback welcome.